pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -203,7 +208,30 @@ func init() {
 }
 
 // Initialize passes a Kubernetes clientBuilder interface to the cloud provider
-func (vs *VSphere) Initialize(clientBuilder controller.ControllerClientBuilder) {}
+func (vs *VSphere) Initialize(clientBuilder controller.ControllerClientBuilder) {
+	if vs.cfg == nil {
+		return
+	}
+
+	// Only on controller node it is required to register listeners.
+	var err error
+	//clientBuilder := clientBuilder
+	client, err := clientBuilder.Client("vSphere-cloud-provider")
+	if err != nil {
+		glog.Errorf("failed to create kubeclient: %v", err)
+      return
```

Comment:
My concern here is that if we fail to initialize for whatever reason, this is not communicated to kubernetes controller manager. Which means, the user would not know if something went wrong during initializing phase. When user tries creating volumes, then things would fail.

How about having a go routine invocation from this method and that routine's only job is to successfully do the initialization? It retries until it is successful. This way, we can recover when things are working fine.

Another approach would be to have a initialized flag in VSphere that is set to true only if the initialization is successful. Every volume operation method will retry the initialization if this flag is set to false.

Reply:
I found a method clientBuilder.ClientOrDie method instead.
So if client can't be created ClientOrDie will make process die. So nothing special will be required.